poloniex/concurrency.py

```diff
@@ -0,0 +1,106 @@
+        self._cond = threading.Condition(threading.Lock())
+        self._initial = self._value = value
+
+    def acquire(self, blocking=True, timeout=None):
+        """Acquire a semaphore, decrementing the internal counter by one.
+        When invoked without arguments: if the internal counter is larger than
+        zero on entry, decrement it by one and return immediately. If it is zero
+        on entry, block, waiting until some other thread has called release() to
+        make it larger than zero. This is done with proper interlocking so that
+        if multiple acquire() calls are blocked, release() will wake exactly one
+        of them up. The implementation may pick one at random, so the order in
+        which blocked threads are awakened should not be relied on. There is no
+        return value in this case.
+        When invoked with blocking set to true, do the same thing as when called
+        without arguments, and return true.
+        When invoked with blocking set to false, do not block. If a call without
+        an argument would block, return false immediately; otherwise, do the
+        same thing as when called without arguments, and return true.
+        When invoked with a timeout other than None, it will block for at
+        most timeout seconds.  If acquire does not complete successfully in
+        that interval, return false.  Return true otherwise.
+        """
+        if not blocking and timeout is not None:
+            raise ValueError("can't specify timeout for non-blocking acquire")
+        rc = False
+        endtime = None
+        with self._cond:
+            while self._value == 0:
+                if not blocking:
+                    break
+                if timeout is not None:
+                    if endtime is None:
+                        endtime = threading._time() + timeout
+                    else:
+                        timeout = endtime - threading._time()
+                        if timeout <= 0:
+                            break
+                self._cond.wait(timeout)
+            else:
+                self._value -= 1
+                rc = True
+        return rc
+
+    __enter__ = acquire
+
+    def release(self):
+        """Release a semaphore, incrementing the internal counter by one.
+        When the counter is zero on entry and another thread is waiting for it
+        to become larger than zero again, wake up that thread.
+        """
+        with self._cond:
+            self._value += 1
+            self._cond.notify()
+
+    def clear(self):
+        """Release the semaphore of all of its bounds, setting the internal
+        counter back to its original bind limit. Notify an equivalent amount
+        of threads that they can run."""
+        with self._cond:
+            to_notify = self._initial - self._value
+            self._value = self._initial
+            self._cond.notify(to_notify)
+
+    def __exit__(self, t, v, tb):
+        self.release()
```

Comment:
Whoops, sorry, I guess my editor didn't add a return here for some reason. I'll rebase and fix when I'm home.